### ISSUE TITLE
Docs: [non-nullable-type-assertion-style] wrong incorrect example

### ISSUE THREAD

**Issue by yeonjuan** [typescript-eslint/typescript-eslint#8646]
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

```ts
const maybe = Math.random() > 0.5 ? '' : undefined;

const definitely = maybe as string;
const alsoDefinitely = <string>maybe;
```
It is the **incorrect"** example in the [documentation](https://typescript-eslint.io/rules/non-nullable-type-assertion-style/#examples) but lint error does not occurred (see [playground](https://typescript-eslint.io/play/#eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6JgeydqYXgEMARklpFStfsmSJo%2BSl1rIiSdFBnQO0SODABfELqA&code=MYewdgzgLgBAtgQwJ4CMCmMC8MCyCoAWAdAE4JgAmIcAFAJQwB8MADEQKwwD8MA5LzABcMAK6U0AMwCWYNBQDcAKEWhIsCpJlSoaADZIs8ZOhgIIMaCRkBzJauindEEABFNYbXoPYAPJZuMiKho8kA)).

A quick look shows that `ts` infers the type of `maybe` as `"" | undefined`. That's why the `as string` assertion is allowed because it casts `maybe`'s type (`"" -> string`)

So, replacing "const" with "let" will cause a lint error.

```ts
let maybe = Math.random() > 0.5 ? '' : undefined;
// maybe's type: string | undefined
const definitely = maybe as string; // Lint error
const alsoDefinitely = <string>maybe; // Lint error
```
 
IMO, the assertions cast the type ("" -> string) which does not do same things with `!`, so the documentation example should be fixed. (otherwise it's a false negative)

### Affected URL(s)

https://typescript-eslint.io/rules/non-nullable-type-assertion-style/#examples

**Comment by auvred** [comment]
Nice catch! This should be fixed in https://github.com/typescript-eslint/typescript-eslint/pull/8497 

https://github.com/typescript-eslint/typescript-eslint/pull/8497/files#diff-c0da64106a63af3d47d001281829950bbc9bd32676d2968e869dbd53899841ed
